### PR TITLE
[Merged by Bors] - feat(Topology/Algebra/PontryaginDual): Local Compactness of the Pontryagin Dual

### PR DESCRIPTION
**Pull request by tb65536**
This PR uses Arzela-Ascoli to prove local compactness of the Pontryagin dual.

---

- [x] depends on: #11334
- [x] depends on: #6844
- [x] depends on: #12002

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by tb65536** [comment]
I'm marking this as awaiting-review in case a lemma other than #12002 would be better for this situation.

File: Mathlib/Topology/Algebra/ContinuousMonoidHom.lean [hunk 1]
```diff
@@ -376,4 +379,75 @@ def compRight {B : Type*} [CommGroup B] [TopologicalSpace B] [TopologicalGroup B
 #align continuous_monoid_hom.comp_right ContinuousMonoidHom.compRight
 #align continuous_add_monoid_hom.comp_right ContinuousAddMonoidHom.compRight
 
+section LocallyCompact
+
+variable {X Y : Type*} [TopologicalSpace X] [Group X] [TopologicalGroup X]
+  [UniformSpace Y] [CommGroup Y] [UniformGroup Y] [T0Space Y] [CompactSpace Y]
+
+theorem locallyCompactSpace_of_equicontinuousAt (U : Set X) (V : Set Y)
```
Comment (j-loreaux) [inline Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:387]: Is there a reason this doesn't work? (I mean, I understand it's not really associated to the Pontryagin Dual, but :shrug:)
```suggestion
@[to_additive]
theorem locallyCompactSpace_of_equicontinuousAt (U : Set X) (V : Set Y)
```

File: Mathlib/Topology/Algebra/ContinuousMonoidHom.lean [hunk 2]
```diff
@@ -376,4 +379,75 @@ def compRight {B : Type*} [CommGroup B] [TopologicalSpace B] [TopologicalGroup B
...
+    have h3 : IsOpen T := isOpen_induced (ContinuousMap.isOpen_setOf_mapsTo hU isOpen_interior)
+    exact h1.mono (interior_maximal h2 h3)
+  exact TopologicalSpace.PositiveCompacts.locallyCompactSpace_of_group
+    ⟨⟨S2, (inducing_toContinuousMap X Y).isCompact_iff.mpr
+      (ArzelaAscoli.isCompact_of_equicontinuous S3 hS4.isCompact h)⟩, hS2⟩
+
+variable [LocallyCompactSpace X]
+
+theorem locallyCompactSpace_of_hasBasis (V : ℕ → Set Y)
+    (hV : ∀ {n x}, x ∈ V n → x * x ∈ V n → x ∈ V (n + 1))
```
Comment (j-loreaux) [inline Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:427]: How related is this to [TopologicalGroup.exists_antitone_basis_nhds_one](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Group/Basic.html#TopologicalGroup.exists_antitone_basis_nhds_one)? I'm just wondering if it's possible to get away with a more natural condition here at all like `FirstCountableTopology`. The answer is likely no, but I thought I'd ask.

**Review by j-loreaux — COMMENTED** [review]
Fingers crossed, but I think it's possible there is a big simplification available here in the case of `FirstCountableTopology X`.

File: Mathlib/Topology/Algebra/ContinuousMonoidHom.lean (hunk 2 above)
Comment (tb65536) [reply Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:427]: Yeah, I think I thought about this possibility, but the condition in docs#TopologicalGroup.exists_antitone_basis_nhds_one isn't enough to get the condition in `hV` (For instance, think about the circle where the first neighborhood is the whole circle and the second neighborhood is the right half of the circle. This satisfies the condition in docs#TopologicalGroup.exists_antitone_basis_nhds_one but not the condition in `hV`). And I think you need a condition like `hV` in order to prove equicontinuity.

Comment (j-loreaux) [reply Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:428]: Can you prove `hV` from first countability? It just feels like these are somehow very closely connected to me.

Comment (tb65536) [reply Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:428]: Hmm, I had thought not, but now I'm not so sure. I guess if you take V(n+1) = Vn cap sq^(-1) (Vn), then this has the property that x in Vn and x^2 in Vn implies x in V(n+1). And you can probably further intersect with the (n+1)st term of an existing countable basis to finish the job?

Comment (tb65536) [reply Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:428]: Ah, the issue is that if you shrink V(n+1) then hV no longer holds. And I wasn't able to find a way around this.

Comment (tb65536) [reply Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:428]: Without some more clever way of modifying the sequence, you would need to find a V0 such that iterating V(n+1) = Vn cap sq^(-1) (Vn) gives a basis.

Comment (j-loreaux) [reply Mathlib/Topology/Algebra/ContinuousMonoidHom.lean:428]: Fair enough. That makes sense. It seemed too good to expect. I didn't see a way to do it either.

**Comment by j-loreaux** [comment]
I haven't checked carefully whether these proofs can be significantly simplified, but I don't think it's too big a problem even if they can be. 

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**